x/cronos/keeper/evm.go

```diff
@@ -0,0 +1,169 @@
+	coinbase, err := k.evmKeeper.GetCoinbaseAddress(ctx)
+	if err != nil {
+		return nil, nil, errors.New("failed to obtain coinbase address")
+	}
+	evm := k.evmKeeper.NewEVM(msg, ethCfg, params, coinbase, nil)
+	ret, err := k.evmKeeper.ApplyMessage(evm, msg, ethCfg, true)
+	if err != nil {
+		return nil, nil, err
+	}
+	k.evmKeeper.CommitCachedContexts()
+	return &msg, ret, nil
+}
+
+// CallModuleCRC20 call a method of ModuleCRC20 contract
+func (k Keeper) CallModuleCRC20(ctx sdk.Context, contract common.Address, method string, args ...interface{}) ([]byte, error) {
+	data, err := types.ModuleCRC20Contract.ABI.Pack(method, args...)
+	if err != nil {
+		return nil, err
+	}
+	_, res, err := k.CallEVM(ctx, &contract, data, big.NewInt(0))
+	if err != nil {
+		return nil, err
+	}
+	if res.Failed() {
+		return nil, fmt.Errorf("call contract failed: %s, %s, %s", contract.Hex(), method, res.Ret)
+	}
+	return res.Ret, nil
+}
+
+// DeployModuleCRC20 deploy an embed erc20 contract
+func (k Keeper) DeployModuleCRC20(ctx sdk.Context, denom string) (common.Address, error) {
+	ctor, err := types.ModuleCRC20Contract.ABI.Pack("", denom, uint8(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+	data := append(types.ModuleCRC20Contract.Bin, ctor...)
+
+	msg, res, err := k.CallEVM(ctx, nil, data, big.NewInt(0))
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	if res.Failed() {
+		return common.Address{}, fmt.Errorf("contract deploy failed: %s", res.Ret)
+	}
+	return crypto.CreateAddress(types.EVMModuleAddress, msg.Nonce()), nil
+}
+
+// ConvertCoinFromNativeToCRC20 convert native token to erc20 token
+func (k Keeper) ConvertCoinFromNativeToCRC20(ctx sdk.Context, sender common.Address, coin sdk.Coin, autoDeploy bool) error {
+	if !types.IsValidDenomToWrap(coin.Denom) {
+		return errors.New("denom is not supported for wrapping")
+	}
+
+	var err error
+	// external contract is returned in preference to auto-deployed ones
+	contract, found := k.GetContractByDenom(ctx, coin.Denom)
+	if !found {
+		if !autoDeploy {
+			return errors.New("no contract found for the denom")
+		}
+		contract, err = k.DeployModuleCRC20(ctx, coin.Denom)
+		if err != nil {
+			return err
+		}
+		k.SetAutoContractForDenom(ctx, coin.Denom, contract)
+	}
+	err = k.bankKeeper.SendCoins(ctx, sdk.AccAddress(sender.Bytes()), sdk.AccAddress(contract.Bytes()), sdk.NewCoins(coin))
+	if err != nil {
+		return err
+	}
+	_, err = k.CallModuleCRC20(ctx, contract, "mint_by_native", sender, coin.Amount.BigInt())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConvertCoinFromCRC20ToNative convert erc20 token to native token
+func (k Keeper) ConvertCoinFromCRC20ToNative(ctx sdk.Context, contract common.Address, receiver common.Address, coin sdk.Coin) error {
+	if !types.IsValidDenomToWrap(coin.Denom) {
+		return errors.New("denom is not supported for wrapping")
+	}
+
+	// validate contract address
+	externalContract, externalFound := k.getExternalContractByDenom(ctx, coin.Denom)
```

Comment:
just use `GetContractByDenom`

Reply:
Here it needs to support both auto and external contracts, user should be able to convert from both.

Reply:
ok I see, maybe in that case u dont need to specify coin but just the amount.

You check if the contract address exists in a reverse map, and you fetch the denom from reverse map.

I am not sure if the denom can be easily found otherwise,  especially if this methods will be called in smart contract.
I imagine smart contract event will be something like 
"__CosmosToNativeToken(amount)" or something like that

Reply:
Actually, I've thought about just store the reversed index, but with the current approach, we can avoid that to save the space in storage by storing the denom in the contract. Since the denom is just `gravity{original_contract_address}`, we are just asking the contract to record their own original address, that looks like a reasonable requirement to me.
Something like this:
```solidity
function original_contract() {
     return 0x...
}
function native_denom() {
    return 'gravity' + original_contract();
}
```

Reply:
so how about removing  `coin sdk.Coin` from the parameter? The denom can be computed from the contract address and it would be easier to use

Reply:
> so how about removing `coin sdk.Coin` from the parameter? The denom can be computed from the contract address and it would be easier to use and replace by amount

But to query denom here, we have to do an EVM call, it would be more efficient to require the contract to pass the denom and just verify it?

Reply:
I though the denom is gravity + contract_address ? 

Oh you mean original contract address in ethereum. Yeah in that case, we could have a reverse mapping `contract_address_in_cronos -> native_token_denom`

It is not a big issue but more to enhance user experience. The main concern is that native_token_denom is not easy to find for users, usually when you want to convert from crc20 to native, you just care about the contract in cronos, you dont care about the ethereum contract. it might be confusing to use.

Reply:
but if it is too bothersome to do, we can have first the denom recorded in SC as you mentioned. In that case we need to modify auto-deployment logic?

It also make external deployment a bit more complex

Reply:
> I though the denom is gravity + contract_address ?
> 
> Oh you mean original contract address in ethereum. Yeah in that case, we could have a reverse mapping `contract_address_in_cronos -> native_token_denom`
> 
> It is not a big issue but more to enhance user experience. The main concern is that native_token_denom is not easy to find for users, usually when you want to convert from crc20 to native, you just care about the contract in cronos, you dont care about the ethereum contract. it might be confusing to use.

hmm, fair enough.

Reply:
done